wolfcrypt/src/evp.c

```diff
@@ -2102,6 +2102,148 @@ WOLFSSL_API int wolfSSL_EVP_PKEY_cmp(const WOLFSSL_EVP_PKEY *a, const WOLFSSL_EV
+    WOLFSSL_BN_CTX* ctx = NULL;
+    WOLFSSL_BIGNUM *num1 = NULL;
+    WOLFSSL_BIGNUM *num2 = NULL;
+    
+    WOLFSSL_ENTER("DH_param_check");
+    
+    ctx = wolfSSL_BN_CTX_new();
+    if (ctx == NULL) {
+        WOLFSSL_MSG("failed to allocate memory");
+        return WOLFSSL_FAILURE;
+    }
+    
+    num1 = wolfSSL_BN_new();
+    num2 = wolfSSL_BN_new();
+    if (num1 == NULL || num2 == NULL) {
+        WOLFSSL_MSG("failed to assign big number");
+        ret = WOLFSSL_FAILURE;
+    }
+    
+    /* prime check */
+    if (ret == WOLFSSL_SUCCESS &&
+        wolfSSL_BN_is_odd(dh_key->p) == 0){
+        WOLFSSL_MSG("dh_key->p is not prime");
+        ret = WOLFSSL_FAILURE;
+    } /* TO DO safe prime check. need BN_rshift1 */
+    
+    /* generator check */
+    if (ret == WOLFSSL_SUCCESS &&
+       (wolfSSL_BN_is_one(dh_key->g) ||
+        wolfSSL_BN_is_negative(dh_key->g) ||
+        wolfSSL_BN_is_zero(dh_key->g))) {
+        WOLFSSL_MSG("dh_key->g is not suitable generator");
+        ret = WOLFSSL_FAILURE;
+    }
+    
+    if (ret == WOLFSSL_SUCCESS &&
+        wolfSSL_BN_cmp(dh_key->p, dh_key->g) <= 0) {
+        WOLFSSL_MSG("dh_key->g is not suitable generator");
+        ret = WOLFSSL_FAILURE;
+    }
+    
+    if (ret == WOLFSSL_SUCCESS &&
+        dh_key->q != NULL)
+    {
+        if (ret == WOLFSSL_SUCCESS &&
+            wolfSSL_BN_mod_exp(num1, dh_key->g, dh_key->q, dh_key->p, ctx) ==
+            WOLFSSL_FAILURE) {
+            WOLFSSL_MSG("BN_mod_exp failed");
+            ret = WOLFSSL_FAILURE;
+        }
+        else
+            if (ret == WOLFSSL_SUCCESS &&
+                wolfSSL_BN_is_one(num1) == WOLFSSL_FAILURE) {
+                WOLFSSL_MSG("dh_key->g is not suitable generator");
+                ret = WOLFSSL_FAILURE;
+            }
+#ifdef WOLFSSL_KEY_GEN
+        /* test if the number q is prime. */
+        if (ret == WOLFSSL_SUCCESS &&
+            (wolfSSL_BN_is_prime_ex(dh_key->q, 64, ctx, NULL) <= 0)) {
+            WOLFSSL_MSG("dh_key->q is not prime or error during check.");
+            ret = WOLFSSL_FAILURE;
+        } /* else TO DO check q div q - 1. need BN_div */
+#endif
+    }
+    
+    /* clean up */
+    wolfSSL_BN_CTX_free(ctx);
+    wolfSSL_BN_free(num1);
+    wolfSSL_BN_free(num2);
+    
+    WOLFSSL_LEAVE("DH_param_check", WOLFSSL_SUCCESS);
+    return ret;
+}
+/**
+ * validate the algorithm parameters
+ * @param ctx  a pointer to WOLFSSL_EVP_PKEY_CTX structure
+ * @return WOLFSSL_SUCCESS on success, otherwise failure
+ */
+int wolfSSL_EVP_PKEY_param_check(WOLFSSL_EVP_PKEY_CTX* ctx)
+{
+    int type;
+    int ret;
+    WOLFSSL_DH* dh_key = NULL;
+    
+    (void)dh_key;
+    
+    /* sanity check */
+    if (ctx == NULL) {
+        return WOLFSSL_FAILURE;
+    }
+    
+    type = wolfSSL_EVP_PKEY_type(wolfSSL_EVP_PKEY_base_id(ctx->pkey));
+    
+    switch(type) {
+        #if !defined(NO_RSA)
+        case EVP_PKEY_RSA:
+            WOLFSSL_MSG("not yet implemented");
```

Comment:
Let's make these messages slightly more specific by mentioning the type in them.  For example "EVP_PKEY_RSA not yet implemented".  Applies to all similar cases below.

Reply:
Thanks for the example. Added type to each cases.